extras/src/main/scala/io/odin/extras/derivation/context.scala

```diff
@@ -0,0 +1,78 @@
+package io.odin.extras.derivation
+
+import io.odin.meta._
+import magnolia.{CaseClass, Magnolia, Param, SealedTrait}
+
+object context {
+
+  type Typeclass[A] = ToContextData[A]
+
+  def combine[A](ctx: CaseClass[Typeclass, A]): Typeclass[A] = value => {
+    if (ctx.isValueClass) {
+      ctx.parameters.headOption.fold[ContextData](ContextMap(Map.empty))(param =>
+          param.typeclass.toContextData(param.dereference(value)))
+    } else {
+      val params = ctx.parameters
+        .filterNot(ContextUtils.isHidden)
+        .collect {
+          case param if ContextUtils.isSecret(param) =>
+            (param.label, ContextStringValue(ContextUtils.SecretPlaceholder))
+
+          // TODO does it make sense to support the @length param?
+
+          case param =>
+            (param.label, param.typeclass.toContextData(param.dereference(value)))
+        }
+        .toMap
+      ContextMap(params)
+    }
+  }
+
+  def dispatch[A](ctx: SealedTrait[Typeclass, A]): Typeclass[A] = value => {
+    ctx.dispatch(value)(sub => sub.typeclass.toContextData(sub.cast(value)))
```

Comment:
I'm wondering if this should add some sort of discriminator field to the result based on which instance `value` was